app/waf/src/controllers/adc.controller.ts

```diff
@@ -732,74 +685,154 @@ export class AdcController extends BaseController {
       });
     }
   }
+
+  private async onboarding(adc: Adc, addon: AddonReqValues): Promise<void> {
+    try {
+      this.logger.debug('start to do onbarding');
+      await this.serialize(adc, {status: AdcState.ONBOARDING});
+
+      let doMgr = await OnboardingManager.instanlize(this.wafapp);
+      let doBody = await doMgr.assembleDo(adc, {onboarding: true});
+      let doId = await doMgr.onboarding(doBody);
+
+      await checkAndWait(() => doMgr.isDone(doId), 240)
+        .then(() =>
+          checkAndWait(() => this.adcStCtr.gotTo(AdcState.ONBOARDED), 240),
+        )
+        .then(() => this.serialize(adc, {status: AdcState.ONBOARDED}));
+    } catch (error) {
+      await this.serialize(adc, {
+        status: AdcState.ONBOARDERR,
+        lastErr: `${AdcState.ONBOARDERR}: ${error}`,
+      });
+    }
+  }
 }
 
-type AddonReqValues = {
+export type AddonReqValues = {
   userToken: string;
   tenantId: string;
 };
 
-class AdcStateCtrlr {
+export class AdcStateCtrlr {
   private states: AdcStateEntry[] = [
     {
       state: AdcState.NEW,
-      check: Promise.resolve(true),
+      check: (ctrl: AdcStateCtrlr) => {
+        return Promise.resolve(true);
+      },
       next: [AdcState.POWERON],
     },
     {
+      failure: AdcState.POWERERR,
       state: AdcState.POWERON,
-      check: this.accessible(),
+      check: this.accessible,
       next: [AdcState.ONBOARDED, AdcState.RECLAIMED],
     },
     {
+      failure: AdcState.ONBOARDERR,
       state: AdcState.ONBOARDED,
-      check: this.onboarded(),
+      check: this.onboarded,
       next: [AdcState.TRUSTED, AdcState.RECLAIMED],
     },
     {
+      failure: AdcState.TRUSTERR,
       state: AdcState.TRUSTED,
-      check: this.trusted(),
-      next: [AdcState.RECLAIMED],
+      check: this.trusted,
+      next: [AdcState.INSTALLED, AdcState.RECLAIMED],
     },
     {
+      failure: AdcState.INSTALLERR,
+      state: AdcState.INSTALLED,
+      check: this.installed,
+      next: [AdcState.RECLAIMED, AdcState.ACTIVE],
+    },
+    {
+      failure: AdcState.RECLAIMERR,
       state: AdcState.RECLAIMED,
-      check: Promise.resolve(this.reclaimed()),
+      check: this.reclaimed,
       next: [AdcState.POWERON],
     },
+    {
+      state: AdcState.ACTIVE,
+      check: (ctrl: AdcStateCtrlr) => {
+        return Promise.resolve(true);
+      },
+      next: [AdcState.RECLAIMED],
+    },
   ];
 
   constructor(private adc: Adc, private addon: AddonReqValues) {}
 
   async readyTo(state: string): Promise<boolean> {
     let stateEntry = this.getStateEntry(this.adc.status);
-    return (await stateEntry.check) && stateEntry.next.includes(state);
+
```

Comment:
pls re-check if   readyTo() logic is correct.  think about the following scenario: e.g. call the {"create": null} or other apis twice?  it seems to me this readyTo() would return true in line 770, which does not seem to be correct.   pls check if running those for the 2nd time does not break anything.

Reply:
That's case which has been considered in adc.controller.ts:

```
420     if (adc.status.endsWith('ING'))
421       throw new HttpErrors.UnprocessableEntity(
422         `Adc status is ' ${adc.status}. Cannot be operated on, please wait for its finish.`,
423       );
```

In the future, this check should also be moved into state controller.

Reply:
It's not about the ING.   the scenario  i meant was:  when the 1st api call  already finished successfully (which means it is not anything ending with ING),   then  the  same api is called for the 2nd time, we  probably want to make sure calling for the 2nd time does not bring anything unexpected.

Reply:
Oops, it's indeed an issue. Done. please review.